data_pipeline/google_analytics/ga_pipeline.py

```diff
@@ -0,0 +1,180 @@
+import os
+import logging
+import json
+from typing import Iterable
+from tempfile import NamedTemporaryFile
+from datetime import datetime
+
+from data_pipeline.google_analytics.ga_config import GoogleAnalyticsConfig
+from data_pipeline.utils.data_store.bq_data_service import (
+    create_or_extend_table_schema,
+    load_file_into_bq
+)
+from data_pipeline.crossref_event_data.etl_crossref_event_data_util import (
+    standardize_field_name
+)
+from data_pipeline.utils.data_pipeline_timestamp import (
+    get_current_timestamp_as_string
+)
+from data_pipeline.google_analytics.etl_state import (
+    update_state
+)
+from data_pipeline.utils.data_store.google_analytics import (
+    GoogleAnalyticsClient
+)
+
+LOGGER = logging.getLogger(__name__)
+
+GA_DATE_RANGE_KEY = "date_range"
+
+
+def transform_response_to_bq_compatible_record(
+        response
+) -> Iterable[dict]:
+    for report in response.get('reports', []):
+        column_header = report.get('columnHeader', {})
+        dimension_headers = column_header.get('dimensions', [])
+        metric_headers = column_header.get('metricHeader', {}).get(
+            'metricHeaderEntries', []
+        )
+
+        for row in report.get('data', {}).get('rows', []):
+            bq_json_formatted_record = {}
+            dimensions = row.get('dimensions', [])
+            date_range_values = row.get('metrics', [])
+
+            for header, dimension in zip(dimension_headers, dimensions):
+                bq_json_formatted_record[
+                    standardize_field_name(header)
+                ] = dimension
+            for i, values in enumerate(date_range_values):
+                bq_json_formatted_record[GA_DATE_RANGE_KEY] = str(i)
+                for metric_header, value in zip(
+                        metric_headers, values.get('values')
+                ):
+                    bq_json_formatted_record[
+                        standardize_field_name(metric_header.get('name'))
+                    ] = value
+
+            yield bq_json_formatted_record
+
+
+def etl_google_analytics(
+        ga_config: GoogleAnalyticsConfig,
+        start_date: datetime,
+        end_date: datetime = None,
+):
+    current_timestamp_as_string = get_current_timestamp_as_string()
+    analytics = GoogleAnalyticsClient()
+    end_date = end_date or datetime.now()
+    from_date = start_date.strftime("%Y-%m-%d")
+    to_date = end_date.strftime("%Y-%m-%d")
+    dimensions = [
+        {
+            'name': dim_name
+        } for dim_name in ga_config.dimensions
+    ]
+    metrics = [
+        {
+            'expression': metric_name
+        } for metric_name in ga_config.metrics
+    ]
+    page_token = None
+    while True:
+
+        response = analytics.get_report(
```

Comment:
I would recommend moving this out to a page generator function. That way this method becomes shorter and it will be easier to see what the stop condition is. And perhaps it's also worth moving the processing of a response to a separate method.

e.g.

```
paged_report_responses = iter_get_report_pages(...)
for paged_report_response in paged_report_responses:
    process_paged_report_response(paged_report_response)
```